test/plugin.test.js

```diff
@@ -0,0 +1,464 @@
+          receivedAmount: '101',
+          maximumAmount: this.plugin._maxPacketAmount.toString()
+        }))
+      }
+    })
+
+    test('throw if amount in prepare puts account over maximum balance',  async () => {
+      const newBalance = this.mockAccount.balance.plus(100)
+      try {
+        await this.plugin._handleCustomData(this.mockAccountAddress, this.mockBtpPacketIlp(this.mockIlpPrepare))
+        throw Error('Test Error: should trigger catch')
+      } catch (e) {
+        expect(e).toEqual(new IlpPacket.Errors.InsufficientLiquidityError(
+          `Insufficient funds, prepared balance is ${formatAmount(newBalance, 'eth')}, above max of ${formatAmount(this.plugin.balance.maximum, 'eth')}`
+        ))
+      }
+    })
+
+    test('throw if no data handler', async () => {
+      this.plugin.balance.maximum = 200 
+      try {
+        await this.plugin._handleCustomData(this.mockAccountAddress, this.mockBtpPacketIlp(this.mockIlpPrepare))
+        throw Error('Test Error: should trigger catch')
+      } catch (e) {
+        expect(e).toEqual(new Error('no request handler registered'))    
+      }
+    })
+
+    describe('on reject response from data handler (from expiry)', () => {
+      beforeEach(async () => {
+        jest.useFakeTimers()
+        this.oldBalance = this.mockAccount.balance
+        this.plugin.balance.maximum = 200
+        this.plugin.registerDataHandler(() => new Promise(resolve => setTimeout(() => resolve(), 100000)))
+        try {
+          this.protocolDataResponse = this.plugin._handleCustomData(this.mockAccountAddress, this.mockBtpPacketIlp(this.mockIlpPrepare))
+          jest.runOnlyPendingTimers()
+          await this.protocolDataResponse
+        }
+        catch (e) {
+          throw e
+        }
+      })
+      
+      test('account balance rolled back', async () => {
+        expect(this.oldBalance).toEqual(this.mockAccount.balance)
+      })
+
+      test('return packet of type reject', async () => {
+        const protocolDataResponse = await this.protocolDataResponse
+        expect(protocolDataResponse[0].data[0]).toBe(IlpPacket.Type.TYPE_ILP_REJECT)
+      })
+    })
+    
+    describe('on fulfill response from data handler', () => {
+      beforeEach(() => {
+        jest.useFakeTimers()
+        this.plugin.registerDataHandler(() => new Promise(resolve => setTimeout(() => resolve(IlpPacket.serializeIlpFulfill(this.mockIlpFulfill)), 1000)))
+        this.plugin.balance.maximum = 200
+      })
+
+      test('account balance updated', async () => {
+        try {
+          const oldBalance = this.mockAccount.balance
+          const protocolDataResponse = this.plugin._handleCustomData(this.mockAccountAddress, this.mockBtpPacketIlp(this.mockIlpPrepare))
+          jest.runOnlyPendingTimers()
+          await protocolDataResponse
+          expect(oldBalance).toEqual(this.mockAccount.balance - this.mockIlpPrepare.amount)
+        } catch (e) {}
```

Comment:
Why are errors being swallowed? that means this test isn't testing anything

Reply:
The intention here (which I did in a couple of other places) was that for longer functions (like handlePrepareResponse), I wanted to test the status of something like a specific like an account's balance. Since it changed so many times over the course of the function, I wanted to capture it during a specific phase -- thus I caught errors which stopped the function early to test the status of the account balance at that point. Do you think it's best to forgoe these tests? I know unit testing logic should not peep inside functions too deeply, but I found this useful in finding a few bugs.